### DISCUSSION TITLE
Sort photos by MetadataDate

### DISCUSSION THREAD

**Discussion by nicosomb** [waschinski/photo-stream#74] [Ideas]
Because I publish old photos on my website https://instantanes.loeuillet.org/, and because I want to display them at the first position of the list (last published, first in the list), I made a change in `photo_filter.rb`, to use MetadataDate (available in Exif values): 

```ruby
      sorted = photos.sort_by { |photo|
          Exiftool.new(photo.path)[:metadata_date]
      }
```

The RSS feed is also generated like that and that's fine for me. 

I didn't have a look to make it cleaner (with a setting in `.env`) because I'm not a ruby expert. 
I just opened this issue to give you the idea 🙂

**Comment by waschinski** [comment]
Hi, you seem to know more about specific Exif values/tags than I do. So I assume `date_time_original` did not work for you? Which Exif values would make sense to get instead of getting all?

Btw I was playing around with setting a specific tag like so `Exiftool.new(photo.path, '-TAG=MetadataDate')` in order to increase performance but then it complains about Exiftool not being installed while it should be due to using the `exiftool_vendored` gem.

**Comment by waschinski** [comment]
Oh and I would probably convert this issue to a discussion if that's okay with you.

**Comment by nicosomb** [comment]
> So I assume date_time_original did not work for you?

In fact, if today I publish a very old photo, the website visitors don't see the new picture quickly. I think it's better to display the last published photo at the top of the list. 

`date_time_original` is very useful, I use it on photo detail to display the date of the photo. 
For example this photo https://instantanes.loeuillet.org/regard It was taken in 2008. The default behavior of photo-stream will display this photo at the end of the list. Not really practical for visitors (and the same for the RSS feed, which is used to tell "hey, there is a new photo, taken in 2008, but it's new on the website").

**Comment by waschinski** [comment]
Ah now I got you! What I still didn't get though is how MetadataDate does help with that unless you explicitly changed it to e.g. today.

And for what you would like to achieve I think even the fallback to the file system `modified_time` won't always work as that might not always be the time the file has been dropped to the `original` folder.

**Comment by nicosomb** [comment]
> how MetadataDate does help with that unless you explicitly changed it to e.g. today.

Here is my workflow to publish photos: https://nicolas.loeuillet.org/billets/2023/03/17/mon-workflow-de-publication-de-photos/ (in french, sorry). 
And the tool I use to convert to `webp` (https://github.com/tmcw/bespoke) seems to define the MetadataDate to `the-date-I-convert-photo-in-webp`.

**Comment by waschinski** [comment]
Okay, so it's bespoke that adds the date data. The thing is, we need to come up with a good general alternative solution to how sorting is currently done. Maybe add an env variable for the exif value to sort by and the default stays `modified_time`? As I would rather not change the current behavior as it would change existing installations on an update.

About your workflow, it was interesting to read how you are doing. The way I am doing it for my gf is by using the docker image (your NAS might be capable of running docker) and mount a folder containing her photos. Then every time I drop new images, the site will be rebuilt automatically and then I manually run the `lftp.sh` script in the container to sync it to her web server.

**Reply by seanwatkins** [reply]
Is it actually modify?

I have 2 files

```
stat 20230618_220027000_iOS.jpg
  File: 20230618_220027000_iOS.jpg
  Size: 596594          Blocks: 1297       IO Block: 1048576 regular file
Device: 97h/151d        Inode: 877223      Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1022/    sean)   Gid: ( 1000/    sean)
Access: 2023-07-02 14:26:27.629905927 -0600
Modify: 2023-06-19 16:15:28.845109224 -0600
Change: 2023-07-01 14:16:05.700449662 -0600
 Birth: -
```
```

stat 'Screenshot 2023-06-30 200844.jpg'
  File: Screenshot 2023-06-30 200844.jpg
  Size: 82639           Blocks: 169        IO Block: 1048576 regular file
Device: 97h/151d        Inode: 900639      Links: 1
Access: (0664/-rw-rw-r--)  Uid: ( 1022/    sean)   Gid: ( 1000/    sean)
Access: 2023-07-02 14:26:25.861914920 -0600
Modify: 2023-07-02 08:37:52.805282384 -0600
Change: 2023-07-02 08:37:52.805282384 -0600

```

So 2nd file, is newer both creation/and modify time than 1st - so it should be first.

For whatever reason, it shows up, way way way down the page?

Any ideas?

**Comment by seanwatkins** [comment]
Interesting running exiftool on both images:



```



'exiftool 20230618_220027000_iOS.jpg
ExifTool Version Number         : 11.88
File Name                       : 20230618_220027000_iOS.jpg
Directory                       : .
File Size                       : 583 kB
File Modification Date/Time     : 2023:06:19 16:15:28-06:00
File Access Date/Time           : 2023:07:02 14:26:27-06:00
File Inode Change Date/Time     : 2023:07:01 14:16:05-06:00
File Permissions                : rw-rw-r--
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.01
Current IPTC Digest             : 3217daede723561f662d2bd94fd332f1
Coded Character Set             : UTF8
Application Record Version      : 2
Time Created                    : 16:00:27
Date Created                    : 2023:06:18
IPTC Digest                     : 3217daede723561f662d2bd94fd332f1
Exif Byte Order                 : Big-endian (Motorola, MM)
Orientation                     : Horizontal (normal)
X Resolution                    : 72
Y Resolution                    : 72
Resolution Unit                 : inches
Y Cb Cr Positioning             : Centered
Exif Version                    : 0221
Date/Time Original              : 2023:06:18 16:00:27
Components Configuration        : Y, Cb, Cr, -
User Comment                    : Screenshot
Flashpix Version                : 0100
Color Space                     : sRGB
Exif Image Width                : 1284
Exif Image Height               : 2778
Scene Capture Type              : Standard
Profile CMM Type                : Apple Computer Inc.
Profile Version                 : 4.0.0
Profile Class                   : Display Device Profile
Color Space Data                : RGB
Profile Connection Space        : XYZ
Profile Date Time               : 2022:01:01 00:00:00
Profile File Signature          : acsp
Primary Platform                : Apple Computer Inc.
CMM Flags                       : Not Embedded, Independent
Device Manufacturer             : Apple Computer Inc.
Device Model                    :
Device Attributes               : Reflective, Glossy, Positive, Color
Rendering Intent                : Perceptual
Connection Space Illuminant     : 0.9642 1 0.82491
Profile Creator                 : Apple Computer Inc.
Profile ID                      : ecfda38e388547c36db4bd4f7ada182f
Profile Description             : Display P3
Profile Copyright               : Copyright Apple Inc., 2022
Media White Point               : 0.96419 1 0.82489
Red Matrix Column               : 0.51512 0.2412 -0.00105
Green Matrix Column             : 0.29198 0.69225 0.04189
Blue Matrix Column              : 0.1571 0.06657 0.78407
Red Tone Reproduction Curve     : (Binary data 32 bytes, use -b option to extract)
Chromatic Adaptation            : 1.04788 0.02292 -0.0502 0.02959 0.99048 -0.01706 -0.00923 0.01508 0.75168
Blue Tone Reproduction Curve    : (Binary data 32 bytes, use -b option to extract)
Green Tone Reproduction Curve   : (Binary data 32 bytes, use -b option to extract)
Image Width                     : 1284
Image Height                    : 1044
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:2:0 (2 2)
Image Size                      : 1284x1044
Megapixels                      : 1.3
Date/Time Created               : 2023:06:18 16:00:27

```



```


exiftool 'Screenshot 2023-06-30 200844.jpg'
ExifTool Version Number         : 11.88
File Name                       : Screenshot 2023-06-30 200844.jpg
Directory                       : .
File Size                       : 81 kB
File Modification Date/Time     : 2023:07:02 08:37:52-06:00
File Access Date/Time           : 2023:07:02 14:26:25-06:00
File Inode Change Date/Time     : 2023:07:02 08:37:52-06:00
File Permissions                : rw-rw-r--
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.01
Resolution Unit                 : cm
X Resolution                    : 37
Y Resolution                    : 37
Image Width                     : 353
Image Height                    : 524
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 353x524
Megapixels                      : 0.185


```
